`ContextScopeListenerTest.testAsyncServletAsyncAfterDoGetExit()` is flaky

**Jetty version(s)**
12.0.x, 12.1.x

**Jetty Environment**
ee10, ee11

**Description**

`testAsyncServletAsyncAfterDoGetExit()` sometimes fails with:

```
java.lang.AssertionError: 
Expected :<[enterScope /initialPath, doGet, enterScope /initialPath, exitScope /initialPath, asyncRunnable, exitScope /initialPath, enterScope /initialPath, asyncDispatch, exitScope /initialPath]>
Actual   :<[enterScope /initialPath, doGet, enterScope /initialPath, exitScope /initialPath, asyncRunnable, enterScope /initialPath, exitScope /initialPath, asyncDispatch, exitScope /initialPath]>
```

See: https://github.com/jetty/jetty.project/pull/12635/checks?check_run_id=34325977342

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java
@@ -589,11 +589,11 @@ public class ContextHandler extends Handler.Wrapper implements Attributes, Alias
     {
         if (super.addEventListener(listener))
         {
-            if (listener instanceof ContextScopeListener)
+            if (listener instanceof ContextScopeListener contextScopeListener)
             {
-                _contextListeners.add((ContextScopeListener)listener);
+                _contextListeners.add(contextScopeListener);
                 if (__context.get() != null)
-                    ((ContextScopeListener)listener).enterScope(__context.get(), null);
+                    contextScopeListener.enterScope(__context.get(), null);
             }
             return true;
         }
@@ -605,9 +605,12 @@ public class ContextHandler extends Handler.Wrapper implements Attributes, Alias
     {
         if (super.removeEventListener(listener))
         {
-            if (listener instanceof ContextScopeListener)
-                _contextListeners.remove(listener);
-
+            if (listener instanceof ContextScopeListener contextScopeListener)
+            {
+                _contextListeners.remove(contextScopeListener);
+                if (__context.get() != null)
+                    contextScopeListener.exitScope(__context.get(), null);
+            }
             return true;
         }
         return false;
@@ -1053,6 +1056,7 @@ public class ContextHandler extends Handler.Wrapper implements Attributes, Alias
             return true;
 
         // Past this point we are calling the downstream handler in scope.
+        Context lastContext = getCurrentContext();
         ClassLoader lastLoader = enterScope(contextRequest);
         ContextResponse contextResponse = wrapResponse(contextRequest, response);
         try
@@ -1068,7 +1072,7 @@ public class ContextHandler extends Handler.Wrapper implements Attributes, Alias
         {
             // We exit scope here, even though handle() is asynchronous,
             // as we have wrapped all our callbacks to re-enter the scope.
-            exitScope(contextRequest, request.getContext(), lastLoader);
+            exitScope(contextRequest, lastContext, lastLoader);
         }
     }
 

--- a/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ContextScopeListenerTest.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ContextScopeListenerTest.java
@@ -201,7 +201,7 @@ public class ContextScopeListenerTest
         ContentResponse response = _client.GET(uri);
         assertThat(response.getStatus(), equalTo(HttpStatus.OK_200));
 
-        Awaitility.waitAtMost(5, TimeUnit.SECONDS).pollInterval(100, TimeUnit.MILLISECONDS).until(() -> _history.size() == 9);
+        Awaitility.waitAtMost(5, TimeUnit.SECONDS).pollInterval(100, TimeUnit.MILLISECONDS).until(() -> _history.size() >= 9);
         assertHistory(
             "enterScope /initialPath",
             "doGet",

